File: 239-sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new LinkedList<>();
        List<Integer> res = new ArrayList<>();

        for(int i = 0 ; i < nums.length ; i++){

          if(!dq.isEmpty() && dq.peekFirst() == i - k){
            dq.pollFirst();
          }

          while(!dq.isEmpty() && nums[dq.peekLast()] < nums[i]){
            dq.pollLast();
          }

          dq.offerLast(i);

          if(i >= k - 1){
            res.add(nums[dq.peekFirst()]);
          }
        }

        int[] r = new int[res.size()];
        int idx = 0;

        for(int num : res){
          r[idx++] = num;
        }

        return r;
    }
}

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = deque()
        n = len(nums)
        res = []

        for i in range(n):

            while q and i-k >= q[0]:
                q.popleft()
            
            while q and nums[i] >= nums[q[-1]]:
                q.pop()
            
            q.append(i)

            if i+1 >= k:
                res.append(nums[q[0]])
        
        return res